client/state/notification-settings/toggle-state.js

```diff
@@ -0,0 +1,67 @@
+/** @format */
+/**
+ * Internal dependencies
+ */
+import { find, get, findIndex } from 'lodash';
+
+const replaceAtIndex = ( array, index, item ) => {
+	const newArray = array.slice();
+	newArray.splice( index, 1, item );
+
+	return newArray;
+};
+
+const toggleInStream = ( streamName, stream, setting ) => ( {
+	[ streamName ]: {
+		...stream,
+		[ setting ]: ! get( stream, setting ),
+	},
+} );
+
+const toggleInDevice = ( devices, deviceId, setting ) => {
+	const device = find( devices, { device_id: parseInt( deviceId, 10 ) } );
+	const deviceIndex = findIndex( devices, device );
+	const deviceSetting = get( device, setting );
+
+	return {
+		devices: replaceAtIndex( devices, deviceIndex, {
```

Comment:
What if `deviceIndex` is -1, i.e., the `device` is not yet in the array? The old code supported the "replace or add" behavior, didn't it?

It seems that it doesn't matter, as the REST API response (and Redux state) has fields for all possible settings for all your blogs and devices. Therefore a field always exists when we toggle it. But I'm not sure and the previous version of the code was more forgiving.